object_gomap_test.go

```diff
@@ -326,3 +328,93 @@ func TestGoMapUnicode(t *testing.T) {
 		t.Fatalf("Unexpected value: %v", res)
 	}
 }
+
+func TestGoMapMemUsage(t *testing.T) {
+	vm := New()
+	tests := []struct {
+		name           string
+		val            *objectGoMapSimple
+		threshold      int
+		expectedMem    uint64
+		expectedNewMem uint64
+		errExpected    error
+	}{
+		{
+			name:      "should account for each key value pair given a non-empty object",
+			threshold: 100,
+			val: &objectGoMapSimple{
+				baseObject: baseObject{
+					val: &Object{runtime: vm},
+				},
+				data: map[string]interface{}{
+					"test0": valueInt(99),
+					"test1": valueInt(99),
+					"test2": valueInt(99),
+					"test3": valueInt(99),
+				},
+			},
+			// overhead + len("testN") + value
+			expectedMem: SizeEmptyStruct + (5+SizeInt)*4,
+			// overhead + len("testN") with string overhead + value
+			expectedNewMem: SizeEmptyStruct + ((5+SizeString)+SizeInt)*4,
+			errExpected:    nil,
+		},
+		{
+			name:      "should account for each key value pair given map with a nil value",
+			threshold: 100,
+			val: &objectGoMapSimple{
+				baseObject: baseObject{
+					val: &Object{runtime: vm},
+				},
+				data: map[string]interface{}{
+					"test": nil,
+				},
+			},
+			// overhead + len("test") + null
+			expectedMem: SizeEmptyStruct + 4 + SizeEmptyStruct,
+			// overhead + len("test") with string overhead + null
+			expectedNewMem: SizeEmptyStruct + (4 + SizeString) + SizeEmptyStruct,
+			errExpected:    nil,
+		},
+		{
+			name:      "should account for key value pair with value over threshold",
+			threshold: 20,
+			val: &objectGoMapSimple{
+				baseObject: baseObject{
+					val: &Object{runtime: vm},
+				},
+				data: map[string]interface{}{
+					"test": []interface{}{
+						valueInt(99),
+						valueInt(99),
+						valueInt(99),
+						valueInt(99),
+					},
+				},
+			},
+			// overhead + len("testN") + value
+			expectedMem: SizeEmptyStruct + 4 + SizeEmptyStruct, // <- why does this work?
+			// overhead + len("testN") with string overhead + value
+			expectedNewMem: SizeEmptyStruct + (4 + SizeString) + SizeEmptyStruct, // <- why does this not account for the inside the slice?
```

Comment:
I'm confused why I'm seeing this memory overhead here. Seems like we need to add additional logic when parsing the map?

Reply:
The slice gets converted to `objectGoSlice` which currently returns `SizeEmptyStruct` which is why you see the overhead. We will need to add a `MemUsage` function for that type as well and remove it from the switch.